src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+// Written in the D programming language.
+/**
+This module provides a composable reference count implementation in the form
+of `__RefCount`.
+
+License: $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Eduard Staniloiu
+
+Source: $(DRUNTIMESRC core/experimental/refcount.d)
+*/
+module core.experimental.refcount;
+
+///
+unittest
+{
+    auto rc = __RefCount.make!__RefCount();
+    assert(rc.isUnique);
+    {
+        auto rc2 = rc; // Construct rc2 by copy construction
+        assert(!rc.isUnique);
+        auto rc3 = __RefCount.make!__RefCount();
+        rc2 = rc3; // Assign rc3 into rc2; rc's ref count drops
+        assert(rc.isUnique);
+        rc3 = rc; // Assign rc into rc3; rc's ref count increases
+        assert(!rc.isUnique);
+        // rc2 and rc3 go out of scope here
+        // rc2 is the last ref to the `__RefCount` created in this block -> gets freed
+    }
+    assert(rc.isUnique);
+}
+
+/**
+A qualified reference counted `struct` that is intended to be composed by
+user-defined types that desire to implement manual memory management by means of
+reference counting.
+
+`__RefCount` was designed to be composed as a field inside a user-defined type.
+The user is responsible to initialize the `__RefCount` in that type's constructors.
+The user will call the `isUnique()` method to decide if this is the last
+reference to the enclosing type so memory can be safely deallocated.
+
+Implementation: The internal implementation of `__RefCount` uses `malloc`/`free`.
+
+Important: The `__RefCount` member must be initialized through a call to its
+constructor before being used.
+*/
+struct __RefCount
+{
+    import core.atomic : atomicOp;
+
+    alias CounterType = uint;
+    private shared CounterType* rc = null;
+
+    /*
+    Perform `rc op val` operation. Always use atomics as the counter is shared.
```

Comment:
Hmm, I guess this is the most controversial part of this PR as this will make it very slow for single-threaded programs.
Can't we separate the overloads, s.t. this is only done when `__RefCount` is actually shared (and thus its `shared` functions are called?).
Alternatively, maybe we need to add a `__TLSRefCount`?

Reply:
Sadly, we can't. With implicit conversion of `immutable(__RefCount)` to `const(__RefCount)` we can't rely on `this` being shared.

The solution @edi33416 was trying out for this used alignment as a marker for `__RefCount`s that were constructed as `shared` (e.g. `immutable`), so after implicit conversion to `const` we can still know it was initially `immutable`. 

However this doesn't work either, as the result of `pure` functions can be implicitly converted to `immutable`. This means that when creating an returning a `__RefCount` from a `pure` function, the non-`shared` constructor ends up being called for a `shared` object, meaning it gets constructed with non-`shared` alignment, and we're right back at square one.

Reply:
I have a whole lot of problems with this conversation. It seems like there are mistakes everywhere.

Reply:
@TurkeyMan Any insights you might have about this particular issue would be very much appreciated.

Reply:
I haven't had time to think on it sorry, as funny as it might seem to some, RC is not actually an urgent issue for me as it has been for the last 10 years >_<  
This is very important, but I need to devote 100% attention to `shared` right now.

Reply:
My instinct is that RC is so important, that `immutable`->`shared` conversion causing us an issue here is satisfactory grounds to re-consider `immutable`->`shared` conversion rules... the problem might not be here, maybe it's there. (I don't know, I haven't thought on it at all, that's just my feeling)

Reply:
I've had the same thought. I think it's definitely worth looking into.